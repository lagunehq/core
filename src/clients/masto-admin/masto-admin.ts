import { AdminAccount } from '../../entities/admin-account';
import { AdminReport } from '../../entities/admin-report';
import { Gateway } from '../../gateway/gateway';
import { available } from '../decorators';
import {
  AdminActionAccountParams,
  AdminFetchAccountParams,
  AdminFetchReportsParams,
} from './params';

/**
 * Mastodon Moderation API client
 */
export class MastoAdmin extends Gateway {
  /**
   * Get accounts
   * @param params parameters
   * @retunr Accounts
   */
  @available({ since: '2.9.1' })
  public fetchAccounts(params?: AdminFetchAccountParams) {
    return this.get<AdminAccount[]>('/api/v1/admin/accounts', params);
  }

  /**
   * Get a specific account
   * @param id ID of the account
   * @return Account
   */
  @available({ since: '2.9.1' })
  public fetchAccount(id: string) {
    return this.get<AdminAccount>(`/api/v1/admin/accounts/${id}`);
  }

  /**
   * Re-enable account
   * @param id ID of the account
   * @return Account
   */
  @available({ since: '2.9.1' })
  public enableAccount(id: string) {
    return this.post<AdminAccount>(`/api/v1/admin/accounts/${id}/enable`);
  }

  /**
   * Approve pending account
   * @param id ID of the account
   * @return Account
   */
  @available({ since: '2.9.1' })
  public approveAccount(id: string) {
    return this.post<AdminAccount>(`/api/v1/admin/accounts/${id}/approve`);
  }

  /**
   * Reject pending account
   * @param id ID of the account
   * @return Account
   */
  @available({ since: '2.9.1' })
  public rejectAccount(id: string) {
    return this.post<AdminAccount>(`/api/v1/admin/accounts/${id}/reject`);
  }

  /**
   * Unsilence account
   * @param id ID of the account
   * @return Account
   */
  @available({ since: '2.9.1' })
  public unsilenceAccount(id: string) {
    return this.post<AdminAccount>(`/api/v1/admin/accounts/${id}/unsilence`);
  }

  /**
   * Unsuspend account
   * @param id ID of the account
   * @return Account
   */
  @available({ since: '2.9.1' })
  public unsuspendAccount(id: string) {
    return this.post<AdminAccount>(`/api/v1/admin/accounts/${id}/unsuspend`);
  }

  /**
   * Get reports
   * @param params params
   * @return Reports
   */
  @available({ since: '2.9.1' })
  public fetchReports(params?: AdminFetchReportsParams) {
    return this.get<AdminReport[]>('/api/v1/admin/reports', params);
  }

  /**
   * Get a specific report
   * @param id ID of the report
   * @return Report
   */
  @available({ since: '2.9.1' })
  public fetchReport(id: string) {
    return this.get<AdminReport>(`/api/v1/admin/reports/${id}`);
  }

  /**
   * Assign report to self
   * @param id ID of the report
   * @return Report
   */
  @available({ since: '2.9.1' })
  public assignReportToSelf(id: string) {
    return this.post<AdminReport>(`/api/v1/admin/reports/${id}/assign_to_self`);
  }

  /**
   * Unassign report from self
   * @param id ID of the report
   * @return Report
   */
  @available({ since: '2.9.1' })
  public unassignReport(id: string) {
    return this.post<AdminReport>(`/api/v1/admin/reports/${id}/unassign`);
  }

  /**
   * Re-open report
   * @param id ID of the report
   * @return Report
   */
  @available({ since: '2.9.1' })
  public reopenReport(id: string) {
    return this.post<AdminReport>(`/api/v1/admin/reports/${id}/reopen`);
  }

  /**
   * Close report
   * @param id ID of the report
   * @return Report
   */
  @available({ since: '2.9.1' })
  public resolveReport(id: string) {
    return this.post<AdminReport>(`/api/v1/admin/reports/${id}/resolve`);
  }

  /**
   * Perform a moderation action on an account
   * @param id ID of the account
   * @param params Params
   * @return Acccount
   */
  @available({ since: '2.9.1' })
  public actionAccount(id: string, params: AdminActionAccountParams) {
    return this.post<AdminAccount>(
      `/api/v1/admin/accounts/${id}/action`,
      params,
    );
  }
}
