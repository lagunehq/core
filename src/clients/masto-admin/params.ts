export interface AdminFetchAccountParams {
  local?: boolean | null;
  remote?: boolean | null;
  by_domain?: string | null;
  active?: boolean | null;
  pending?: boolean | null;
  disabled?: boolean | null;
  silenced?: boolean | null;
  suspended?: boolean | null;
  usename?: string | null;
  display_name?: string | null;
  email?: string | null;
  ip?: string | null;
  staff?: boolean | null;
}

export interface AdminFetchReportsParams {
  resolved?: boolean | null;
  account_id?: string | null;
  target_account_id?: string | null;
}

export type AccountActionType = 'none' | 'disable' | 'silence' | 'suspend';

export interface AdminActionAccountParams {
  type: AccountActionType;
  report_id: string;
  warning_preset_id?: string | null;
  text?: string | null;
  send_email_notification?: boolean | null;
}
