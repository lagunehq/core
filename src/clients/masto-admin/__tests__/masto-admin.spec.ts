// @ts-ignore
import axios from 'axios';
import { MastoAdmin } from '../masto-admin';

jest.mock('axios');

describe('MastoAdmin', () => {
  const masto = new MastoAdmin({
    uri: 'https://example.com',
    version: '99.99.9',
    streamingApiUrl: 'wss://example.com/stream',
  });

  beforeAll(async () => {
    (axios.request as jest.Mock).mockResolvedValue({
      headers: {
        link: '<https://example.com/next>; rel="next"',
      },
      data: {},
    });
  });

  beforeEach(async () => {
    masto.version = '99.99.9'; // avoid version error
    (axios.request as jest.Mock).mockClear();
  });

  test('fetchAccounts', async () => {
    await masto.fetchAccounts();
    expect(axios.request as jest.Mock).toBeCalledTimes(1);
    expect(axios.request as jest.Mock).toMatchSnapshot();
  });

  test('fetchAccount', async () => {
    await masto.fetchAccount('123');
    expect(axios.request as jest.Mock).toBeCalledTimes(1);
    expect(axios.request as jest.Mock).toMatchSnapshot();
  });

  test('enableAccount', async () => {
    await masto.enableAccount('123');
    expect(axios.request as jest.Mock).toBeCalledTimes(1);
    expect(axios.request as jest.Mock).toMatchSnapshot();
  });

  test('approveAccount', async () => {
    await masto.approveAccount('123');
    expect(axios.request as jest.Mock).toBeCalledTimes(1);
    expect(axios.request as jest.Mock).toMatchSnapshot();
  });

  test('rejectAccount', async () => {
    await masto.rejectAccount('123');
    expect(axios.request as jest.Mock).toBeCalledTimes(1);
    expect(axios.request as jest.Mock).toMatchSnapshot();
  });

  test('unsilenceAccount', async () => {
    await masto.unsilenceAccount('123');
    expect(axios.request as jest.Mock).toBeCalledTimes(1);
    expect(axios.request as jest.Mock).toMatchSnapshot();
  });

  test('unsuspendAccount', async () => {
    await masto.unsuspendAccount('123');
    expect(axios.request as jest.Mock).toBeCalledTimes(1);
    expect(axios.request as jest.Mock).toMatchSnapshot();
  });

  test('fetchReports', async () => {
    await masto.fetchReports();
    expect(axios.request as jest.Mock).toBeCalledTimes(1);
    expect(axios.request as jest.Mock).toMatchSnapshot();
  });

  test('fetchReport', async () => {
    await masto.fetchReport('123');
    expect(axios.request as jest.Mock).toBeCalledTimes(1);
    expect(axios.request as jest.Mock).toMatchSnapshot();
  });

  test('assignReportToSelf', async () => {
    await masto.assignReportToSelf('123');
    expect(axios.request as jest.Mock).toBeCalledTimes(1);
    expect(axios.request as jest.Mock).toMatchSnapshot();
  });

  test('unassignReport', async () => {
    await masto.unassignReport('123');
    expect(axios.request as jest.Mock).toBeCalledTimes(1);
    expect(axios.request as jest.Mock).toMatchSnapshot();
  });

  test('reopenReport', async () => {
    await masto.reopenReport('123');
    expect(axios.request as jest.Mock).toBeCalledTimes(1);
    expect(axios.request as jest.Mock).toMatchSnapshot();
  });

  test('resolveReport', async () => {
    await masto.resolveReport('123');
    expect(axios.request as jest.Mock).toBeCalledTimes(1);
    expect(axios.request as jest.Mock).toMatchSnapshot();
  });

  test('actionAccount', async () => {
    await masto.actionAccount('123', {
      type: 'disable',
      report_id: '123123',
      warning_preset_id: '123123',
      text: 'Your account have been disabled',
      send_email_notification: false,
    });
    expect(axios.request as jest.Mock).toBeCalledTimes(1);
    expect(axios.request as jest.Mock).toMatchSnapshot();
  });
});
