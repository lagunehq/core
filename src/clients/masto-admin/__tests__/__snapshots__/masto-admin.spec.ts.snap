// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MastoAdmin actionAccount 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": "{\\"type\\":\\"disable\\",\\"report_id\\":\\"123123\\",\\"warning_preset_id\\":\\"123123\\",\\"text\\":\\"Your account have been disabled\\",\\"send_email_notification\\":false}",
        "headers": Object {
          "Content-Type": "application/json",
        },
        "method": "POST",
        "transformResponse": Array [
          [Function],
        ],
        "url": "https://example.com/api/v1/admin/accounts/123/action",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`MastoAdmin approveAccount 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": "{}",
        "headers": Object {
          "Content-Type": "application/json",
        },
        "method": "POST",
        "transformResponse": Array [
          [Function],
        ],
        "url": "https://example.com/api/v1/admin/accounts/123/approve",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`MastoAdmin assignReportToSelf 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": "{}",
        "headers": Object {
          "Content-Type": "application/json",
        },
        "method": "POST",
        "transformResponse": Array [
          [Function],
        ],
        "url": "https://example.com/api/v1/admin/reports/123/assign_to_self",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`MastoAdmin enableAccount 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": "{}",
        "headers": Object {
          "Content-Type": "application/json",
        },
        "method": "POST",
        "transformResponse": Array [
          [Function],
        ],
        "url": "https://example.com/api/v1/admin/accounts/123/enable",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`MastoAdmin fetchAccount 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": undefined,
        "headers": Object {
          "Content-Type": "application/json",
        },
        "method": "GET",
        "params": Object {},
        "transformResponse": Array [
          [Function],
        ],
        "url": "https://example.com/api/v1/admin/accounts/123",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`MastoAdmin fetchAccounts 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": undefined,
        "headers": Object {
          "Content-Type": "application/json",
        },
        "method": "GET",
        "params": Object {},
        "transformResponse": Array [
          [Function],
        ],
        "url": "https://example.com/api/v1/admin/accounts",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`MastoAdmin fetchReport 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": undefined,
        "headers": Object {
          "Content-Type": "application/json",
        },
        "method": "GET",
        "params": Object {},
        "transformResponse": Array [
          [Function],
        ],
        "url": "https://example.com/api/v1/admin/reports/123",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`MastoAdmin fetchReports 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": undefined,
        "headers": Object {
          "Content-Type": "application/json",
        },
        "method": "GET",
        "params": Object {},
        "transformResponse": Array [
          [Function],
        ],
        "url": "https://example.com/api/v1/admin/reports",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`MastoAdmin rejectAccount 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": "{}",
        "headers": Object {
          "Content-Type": "application/json",
        },
        "method": "POST",
        "transformResponse": Array [
          [Function],
        ],
        "url": "https://example.com/api/v1/admin/accounts/123/reject",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`MastoAdmin reopenReport 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": "{}",
        "headers": Object {
          "Content-Type": "application/json",
        },
        "method": "POST",
        "transformResponse": Array [
          [Function],
        ],
        "url": "https://example.com/api/v1/admin/reports/123/reopen",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`MastoAdmin resolveReport 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": "{}",
        "headers": Object {
          "Content-Type": "application/json",
        },
        "method": "POST",
        "transformResponse": Array [
          [Function],
        ],
        "url": "https://example.com/api/v1/admin/reports/123/resolve",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`MastoAdmin unassignReport 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": "{}",
        "headers": Object {
          "Content-Type": "application/json",
        },
        "method": "POST",
        "transformResponse": Array [
          [Function],
        ],
        "url": "https://example.com/api/v1/admin/reports/123/unassign",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`MastoAdmin unsilenceAccount 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": "{}",
        "headers": Object {
          "Content-Type": "application/json",
        },
        "method": "POST",
        "transformResponse": Array [
          [Function],
        ],
        "url": "https://example.com/api/v1/admin/accounts/123/unsilence",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`MastoAdmin unsuspendAccount 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": "{}",
        "headers": Object {
          "Content-Type": "application/json",
        },
        "method": "POST",
        "transformResponse": Array [
          [Function],
        ],
        "url": "https://example.com/api/v1/admin/accounts/123/unsuspend",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;
